pkg/services/automation/generator.go

```diff
@@ -0,0 +1,494 @@
+	)
+
+	if wego.SourceType(app.Spec.SourceType) == wego.SourceType(wego.SourceTypeGit) {
+		appRepoUrl, err = gitproviders.NewRepoURL(app.Spec.URL)
+		if err != nil {
+			return models.Application{}, err
+		}
+	} else {
+		helmRepoUrl = app.Spec.URL
+	}
+
+	if models.IsExternalConfigUrl(app.Spec.ConfigURL) {
+		configRepoUrl, err = gitproviders.NewRepoURL(app.Spec.ConfigURL)
+		if err != nil {
+			return models.Application{}, err
+		}
+	}
+
+	return models.Application{
+		Name:                app.Name,
+		Namespace:           app.Namespace,
+		GitSourceURL:        appRepoUrl,
+		HelmSourceURL:       helmRepoUrl,
+		ConfigURL:           configRepoUrl,
+		Branch:              app.Spec.Branch,
+		Path:                app.Spec.Path,
+		HelmTargetNamespace: app.Spec.HelmTargetNamespace,
+		SourceType:          models.SourceType(app.Spec.SourceType),
+		AutomationType:      models.AutomationType(app.Spec.DeploymentType),
+	}, nil
+}
+
+func AppToWegoApp(app models.Application) wego.Application {
+	sourceUrl := app.GitSourceURL.String()
+
+	if app.SourceType == models.SourceTypeHelm {
+		sourceUrl = app.HelmSourceURL
+	}
+
+	gvk := wego.GroupVersion.WithKind(wego.ApplicationKind)
+	wegoApp := wego.Application{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       gvk.Kind,
+			APIVersion: gvk.GroupVersion().String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      app.Name,
+			Namespace: app.Namespace,
+		},
+		Spec: wego.ApplicationSpec{
+			ConfigURL:           app.ConfigURL.String(),
+			Branch:              app.Branch,
+			URL:                 sourceUrl,
+			Path:                app.Path,
+			DeploymentType:      wego.DeploymentType(string(app.AutomationType)),
+			SourceType:          wego.SourceType(string(app.SourceType)),
+			HelmTargetNamespace: app.HelmTargetNamespace,
+		},
+	}
+
+	return wegoApp
+}
+
+// Operations to extract useful information from an Application
+
+const automationRoot = ".weave-gitops"
+
+func AppYamlDir(a models.Application) string {
+	return filepath.Join(automationRoot, "apps", a.Name)
+}
+
+func AppYamlPath(a models.Application) string {
+	return filepath.Join(AppYamlDir(a), "app.yaml")
+}
+
+func AppAutomationSourcePath(a models.Application) string {
+	return filepath.Join(AppYamlDir(a), fmt.Sprintf("%s-gitops-source.yaml", a.Name))
+}
+
+func AppAutomationDeployPath(a models.Application) string {
+	return filepath.Join(AppYamlDir(a), fmt.Sprintf("%s-gitops-deploy.yaml", a.Name))
+}
+
+func AppAutomationKustomizePath(a models.Application) string {
+	return filepath.Join(AppYamlDir(a), "kustomization.yaml")
+}
+
+func AutomationUserKustomizePath(clusterName string) string {
+	return filepath.Join(automationRoot, "clusters", clusterName, "user", "kustomization.yaml")
+}
+
+func AppSourceName(a models.Application) string {
+	return a.Name
+}
+
+func AppDeployName(a models.Application) string {
+	return a.Name
+}
+
+func AppResourceName(a models.Application) string {
+	return a.Name
+}
+
+type GeneratedSecretName string
```

Comment:
nit: I find the code easier to read when types are defined at the top of the file or at a minimum before their first use.  `GeneratedSecretName` is used in `getAppSecretRef()` at the top of the file

Reply:
:heavy_check_mark: